cpu/nrf52/periph/spi.c

```diff
@@ -0,0 +1,240 @@
+    NRF_PPI->CH[spi_config[bus].ppi].TEP = (uint32_t)&dev(bus)->TASKS_STOP;
+}
+
+static void _enable_workaround(spi_t bus)
+{
+    /**
+     * The spim instance cannot be stopped mid-byte, so it will finish
+     * transmitting the first byte and then stop. Effectively ensuring
+     * that only 1 byte is transmitted.
+     */
+    NRF_PPI->CHENSET = 1U << spi_config[bus].ppi;
+    gpio_irq_enable(spi_config[bus].sclk);
+}
+
+static void _clear_workaround(spi_t bus)
+{
+    NRF_PPI->CHENCLR = 1U << spi_config[bus].ppi;
+}
+#endif
+
+void spi_init(spi_t bus)
+{
+    assert(bus < SPI_NUMOF);
+
+    /* initialize mutex */
+    mutex_init(&locks[bus]);
+    mutex_init(&busy[bus]);
+    mutex_lock(&busy[bus]);
+    /* initialize pins */
+    spi_init_pins(bus);
+}
+
+void spi_init_pins(spi_t bus)
+{
+    gpio_init(spi_config[bus].sclk, GPIO_OUT);
+    gpio_init(spi_config[bus].mosi, GPIO_OUT);
+    gpio_init(spi_config[bus].miso, GPIO_IN);
+    /* select pins for the SPI device */
+    SPI_SCKSEL  = spi_config[bus].sclk;
+    SPI_MOSISEL = spi_config[bus].mosi;
+    SPI_MISOSEL = spi_config[bus].miso;
+#ifdef CPU_MODEL_NRF52832XXAA
+    _setup_workaround_for_ftpan_58(bus);
+#endif
+    spi_twi_irq_register_spi(dev(bus), spi_isr_handler, (void*)bus);
+}
+
+int spi_acquire(spi_t bus, spi_cs_t cs, spi_mode_t mode, spi_clk_t clk)
+{
+    (void) cs;
+
+    mutex_lock(&locks[bus]);
+    /* configure bus */
+    dev(bus)->CONFIG = mode;
+    dev(bus)->FREQUENCY = clk;
+    /* enable the bus */
+    dev(bus)->ENABLE = SPIM_ENABLE_ENABLE_Enabled;
+
+    return SPI_OK;
+}
+
+void spi_release(spi_t bus)
+{
+    /* power off everything */
+    dev(bus)->ENABLE = 0;
+    mutex_unlock(&locks[bus]);
+}
+
+static void _transfer(spi_t bus, const uint8_t *out_buf, uint8_t *in_buf, uint8_t transfer_len)
+{
+    uint8_t out_len = (out_buf) ? transfer_len : 0;
+    uint8_t in_len = (in_buf) ? transfer_len : 0;
+    const uint8_t *out_mbuf = out_buf;
+    /**
+     * Copy the out buffer in case it resides in flash, EasyDMA only works from
+     * RAM
+     */
+    if ((out_buf < (uint8_t*)SPI_CPU_FLASH_END) && (out_len)) {
+        memcpy(_mbuf[bus], out_buf, out_len);
+        out_mbuf = _mbuf[bus];
+    }
+    dev(bus)->TXD.PTR = (uint32_t)out_mbuf;
+    dev(bus)->RXD.PTR = (uint32_t)in_buf;
+
+    dev(bus)->TXD.MAXCNT = out_len;
+    dev(bus)->RXD.MAXCNT = in_len;
+
+    /* clear any spurious END events */
+    dev(bus)->EVENTS_END = 0;
+    dev(bus)->TASKS_START = 1;
+}
+
+void spi_transfer_bytes(spi_t bus, spi_cs_t cs, bool cont,
+                        const void *out, void *in, size_t len)
+{
+    const uint8_t *out_buf = out;
+    uint8_t *in_buf = in;
+
+    assert(out_buf || in_buf);
+
+    if (cs != SPI_CS_UNDEF) {
+        gpio_clear((gpio_t)cs);
+    }
+
+    /* Enable the workaround when the length is only 1 byte */
+#ifdef CPU_MODEL_NRF52832XXAA
+    size_t _len = len;
+    if (_len == 1) {
```

Comment:
That doesn't work in this case because `spi_conf_t::ppi` is only available when the workaround is compiled in. Unless you want me to add a dummy function for `_enable_workaround`

Reply:
> Unless you want me to add a dummy function for `_enable_workaround`

I'd say it's just personal preference, but with a dummy function you'll keep all the ifdefs contained in one place.

But I'm undecided if that makes the code easier to read, there are not that many in the first place - you can decide what's more pretty.